internal/charmstore/debug.go

```diff
@@ -14,60 +14,304 @@ import (
+
+	// Retrieve the number of base entities.
+	baseNum, err := h.store_.DB.BaseEntities().Count()
+	if err != nil {
+		result.Value = "Cannot count base entities: " + err.Error()
+		return resultKey, result
+	}
+
+	// Retrieve the number of entities.
+	num, err := h.store_.DB.Entities().Count()
+	if err != nil {
+		result.Value = "Cannot count entities for consistency check: " + err.Error()
+		return resultKey, result
+	}
+
+	result.Value = fmt.Sprintf("count: %d", baseNum)
+	result.Passed = num >= baseNum
+	return resultKey, result
+}
+
+func (h *debugReqHandler) checkLogs(
+	resultKey, resultName string,
+	logType mongodoc.LogType,
+	startPrefix, endPrefix string,
+) debugstatus.CheckerFunc {
+	return func() (key string, result debugstatus.CheckResult) {
+		result.Name = resultName
+		start, end, err := h.findTimesInLogs(logType, startPrefix, endPrefix)
+		if err != nil {
+			result.Value = err.Error()
+			return resultKey, result
 		}
-		c := make(chan result)
-		for name, check := range checks {
-			name, check := name, check
-			go func() {
-				c <- result{name: name, err: check()}
-			}()
+		result.Value = fmt.Sprintf("started: %s, completed: %s", start.Format(time.RFC3339), end.Format(time.RFC3339))
+		result.Passed = !(start.IsZero() || end.IsZero())
+		return resultKey, result
+	}
+}
+
+// findTimesInLogs goes through logs in reverse order finding when the start and
+// end messages were last added.
+func (h *debugReqHandler) findTimesInLogs(logType mongodoc.LogType, startPrefix, endPrefix string) (start, end time.Time, err error) {
+	var log mongodoc.Log
+	iter := h.store_.DB.Logs().
+		Find(bson.D{
+		{"level", mongodoc.InfoLevel},
+		{"type", logType},
+	}).Sort("-time", "-id").Iter()
+	for iter.Next(&log) {
+		var msg string
+		if err := json.Unmarshal(log.Data, &msg); err != nil {
+			// an error here probably means the log isn't in the form we are looking for.
+			continue
 		}
-		results := make(map[string]string, n)
-		var failed bool
-		for ; n > 0; n-- {
-			res := <-c
-			if res.err == nil {
-				results[res.name] = "OK"
-			} else {
-				failed = true
-				results[res.name] = res.err.Error()
-			}
+		if start.IsZero() && strings.HasPrefix(msg, startPrefix) {
+			start = log.Time
 		}
-		if failed {
-			keys := make([]string, 0, len(results))
-			for k := range results {
-				keys = append(keys, k)
-			}
-			sort.Strings(keys)
-			msgs := make([]string, len(results))
-			for i, k := range keys {
-				msgs[i] = fmt.Sprintf("[%s: %s]", k, results[k])
-			}
-			return nil, errgo.Newf("check failure: %s", strings.Join(msgs, " "))
+		if end.IsZero() && strings.HasPrefix(msg, endPrefix) {
+			end = log.Time
+		}
+		if !start.IsZero() && !end.IsZero() {
+			break
+		}
+	}
+	if err = iter.Close(); err != nil {
+		return time.Time{}, time.Time{}, errgo.Notef(err, "Cannot query logs")
+	}
+	return
+}
+
+type debugCheckRequest struct {
+	httprequest.Route `httprequest:"GET /debug/check"`
+}
+
+// GET /debug/check.
```

Comment:
It would be nice to have a comment on this saying how it
differs from /debug/status and why.